PowerToys Run hashing functions

IHorvalds opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR adds a plugin to the Run utility for generating GUIDs, hashing and base64 encoding strings. It also should make it fairly easy to add other value generators in the future.

~I have not written any tests for this yet and only tested the functionality manually.~ There are unit tests for the GUID generator and for the input parser.

It supports the following hashing functions:
- MD5
- SHA1
- SHA256
- SHA384
- SHA512

And GUID/UUID versions 1, 3, 4 and 5. I skipped version 2, but if anyone thinks it would be helpful, it could be added.

At the moment, usage is not very obvious from the outset. 
I couldn't think of a way to make it very obvious what hashing functions are supported.
Also, GUID v3 and v5 take exactly 2 parameters - a namespace and a name, separated by space. There are predefined values for the namespace (`ns:DNS`, `ns:URL`, `ns:OID`, `ns:X500`), but any valid GUID would work. I couldn't think of a way to make this obvious either. Suggestions and opinions are welcome!

Usage:
- Generating a random GUID/UUID. Both `guid` and `uuid` keywords are valid and will do the same thing. The query is not case sensitive.
![guid_basic_ss](https://github.com/microsoft/PowerToys/assets/22452178/685f68d4-9826-48d6-99c4-b1170e43cc61)
![uuid_basic_ss](https://github.com/microsoft/PowerToys/assets/22452178/4e55cd77-cdba-4934-8ec1-d5c25e1ead87)

- Generating a time based GUID/UUID (version 1)
![time_based_guid](https://github.com/microsoft/PowerToys/assets/22452178/fb049d31-7796-4660-b7e2-40c2c634ccbe)


- Generating a v3 GUID/UUID for the DNS predefined namespace and name `www.microsoft.com`. Notice the `v` after `guid/uuid` is optional. `guid4` === `guidv4`
![guid_v3_ss](https://github.com/microsoft/PowerToys/assets/22452178/f4242315-a92d-4fce-b7a8-5ba1e3f167f5)


- Generating a v3 GUID/UUID for a random namespace and name `www.microsoft.com`
![Short screen recording](https://github.com/microsoft/PowerToys/assets/22452178/7242f200-99f1-47dc-897b-545ce07c2298)

- Calculating the MD5 hash for 'abc'
![md5_ss](https://github.com/microsoft/PowerToys/assets/22452178/1f1fef0e-b0e1-4cf0-afb7-dc4b6e7c9118)


- Calculating the SHA1 hash for 'abc'.
![sha1_ss](https://github.com/microsoft/PowerToys/assets/22452178/c3c817b7-5dd2-4183-bb80-32be9b78bcba)

- Base64 encoding the string 'abcd'
![image](https://github.com/microsoft/PowerToys/assets/22452178/975a18ec-8ca9-46d4-9746-a67021c19f8f)


<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #24931
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [x] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [x] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [x] **Documentation updated:** https://github.com/MicrosoftDocs/windows-dev-docs/pull/4509

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

IHorvalds committed:
Add hashing plugin for Run

IHorvalds committed:
Merge branch 'main' of https://github.com/IHorvalds/PowerToys into powertoys_run_hashing_functions

IHorvalds committed:
Cleanup logic for hasher plugin

- The IComputeRequest interface should make it easier to implement new
  generators in the future
- The GUID generator can now generate all versions of GUID (a.k.a. UUID)
- The input for the hash functions is not quite right. The
  Wox.Plugin.Query class doesn't actually have a way to ge the raw query
  as given by the user. The issue is with multiple spaces. An input like
  "a a  a    a    a" would only be accessible as "a a a a a" using the
  Query class.
  So for now, hashing only works correctly if the input
  doesn't contain multiple consecutive spaces.
- Need a way to make clear the usage for generating GUIDv3 and v5, since
  they take 2 parameters. There are defaults, but they aren't very
  clear.

htcfreek reviewed (commented):
I know that this is wip. But I suggest to change the name and description. As you wrote in the PR details it is possible to add other generators later.

A screenshot would be great. Especially to understand and see the syntax.

IHorvalds committed:
Change plugin name to ValueGenerator

IHorvalds commented on src/modules/launcher/Plugins/Community.PowerToys.Run.Plugin.Hasher/plugin.json:
```diff
@@ -0,0 +1,14 @@
+  "ID": "a26b1bb4dbd911edafa10242ac120002",
+  "ActionKeyword": "#",
+  "isGlobal": true,
+  "Name": "Hasher",
```
I forgot to change that before making the draft. Thanks for the suggestion! Updated. :)

IHorvalds commented:
@htcfreek I changed the PR summary and added a few screenshots and renamed the plugin to ValueGenerator. Hopefully the usage is slightly clearer now, but I should definitely write some usage documentation similar to what is on the [Run page](https://learn.microsoft.com/en-us/windows/powertoys/run).
I'll make a PR for the docs repo when we mark this draft for review.

IHorvalds committed:
Clean up error handling for the input parser

htcfreek commented:
Two suggestions:
1. Move the result type (guid, uuid, sha1, ..) to the sub title. Having the result twice (subtitle too) makes no sense.
2. Adding a short description for the guid types in the subtitle (e.g. "time based calculation").
3. Adding a copy button and a copy action on pressing Enter key like we have for Calculator results.

IHorvalds commented:
I'll do 1 and 2 now. About 3, it already does that. Clicking on the result or pressing Enter will copy the result to the clipboard.

IHorvalds committed:
Add result type and description to subtitle

IHorvalds commented:
I will update the summary tommorrow with new screenshots.

IHorvalds committed:
Change the icons

IHorvalds committed:
Merge branch 'main' of https://github.com/IHorvalds/PowerToys into powertoys_run_hashing_functions

IHorvalds committed:
Merge branch 'microsoft:main' into powertoys_run_hashing_functions

IHorvalds committed:
Merge branch 'main' of https://github.com/IHorvalds/PowerToys into powertoys_run_hashing_functions

IHorvalds committed:
Add Base64 encoding and unit tests

This commit adds Base64 encoding as a utility of the value generator
plugin. The command is `# base64 ***input***`.

Also added unit tests for the UUID/GUID generator and for the input
parser. I don't think tests are necessary for the hashing functions
or for the base64 encoder, since those were part of the the system
libraries.

I'll open a PR for the documentation tomorrow and mark this draft
for review.

IHorvalds committed:
Excluded UUIDv2

IHorvalds committed:
Change icons

IHorvalds committed:
Add RawUserQuery to Wox.Plugin.Query

Getting the RawUserQuery is necessary to be able to handle queries like
`# md5 a    a`, where the intent is to get the hash for `a    a`.
The existing `RawQuery` removes consecutive whitespaces and there was no
other way of getting the request as entered by the user.

IHorvalds committed:
Add ValueGenerator plugin to installer

Also add the unit tests for the plugin to the pipeline.

IHorvalds committed:
Small cleanup

htcfreek commented:
@IHorvalds 
I had no time to review it yet.

Please fix the typos and add the false-positoves to the [expect list](https://github.com/microsoft/PowerToys/blob/main/.github/actions/spell-check/expect.txt).

IHorvalds committed:
Merge branch 'main' of https://github.com/IHorvalds/PowerToys into powertoys_run_hashing_functions

IHorvalds committed:
Fix spelling

IHorvalds committed:
Fix spelling again

IHorvalds committed:
Spell check for guiddata

htcfreek reviewed (changes requested):
@IHorvalds 
Sory for the delay. Is was a bit busy the last time.

I found a few things that could be improved:
1. Please add the plugin project as dependency in the PowerLauncher project.
2. I think it makes sense to list the supported SHA variants in the error message.
![image](https://github.com/microsoft/PowerToys/assets/61519853/ad07bd05-9839-4c82-82ed-2237af14871a)
4. For all the hashing functions: Hashing space makes no sense. You should wait for the text parameter.
5. For GUIDv3 and GUIDv5: The namespace should support lower case notation.
![image](https://github.com/microsoft/PowerToys/assets/61519853/1817805e-76e4-427c-bb62-25b35f04721e)
6. I see no need to add user messages (error messages) to the debug log as they are shown in the ui. I had 239 entries from a few searches.
7. The log entry "Value Generator: No arguments passed to hasher plugin" is not helpful. First the wrong plugin name and secondly you don't know what the query/requested value type is.
8. The log entry "Value Generator: Invalid Query" should show the entered query. Otherwise it is not helpful.

But in general the plugin works as expected.

IHorvalds commented:
@htcfreek
Thank you for the feedback! I hope I didn't rush you or anything!

The changes seem reasonable. I'll make them now and hopefully it will be done in a couple of hours.

IHorvalds committed:
More fixes

This commit adds the dev docs explaining the classes in the new plugin.
It also fixes the following issues:
1. ValueGenerator was not a dependency for the PowerLauncher project
2. The error message for an invalid SHA variant now displays the
   supported SHA variants
3. Hash requests now wait for a string to hash (i.e. no longer hash an
   empty string)
4. GUID v3 and v5 namespace aliases allow lowercase notation
5. Unnecessary debug logs
6. An empty query will now just log "Empty request"
7. An invalid query will also log user query

IHorvalds committed:
Spell check fix again

htcfreek commented:
> Although, I'm not sure how to add the plugin project as a dependency for the PowerLauncher project. Never mind, found it.

@IHorvalds 
Fyi, this is required to debug the code with "PowerLauncher" as startup project.

IHorvalds commented:
Before adding it, I just built the plugin separately and then started PowerLauncher with the debug configuration.

htcfreek reviewed (approved):
Looks good to me.

_Small thing: Do/Should we show a list of supported Guid version if the queryis wrong?_

IHorvalds commented:
Currently, the plugin doesn't show a list of supported GUID versions. It just says "Undefined GUID version", even for version 2, which is defined, but I didn't implement it here.
But yes, I think it should show the supported versions and have a separate error message for version 2.

Perhaps like this?
![image](https://github.com/microsoft/PowerToys/assets/22452178/3ae7ac85-3eee-489a-b998-6a532f426024)
![image](https://github.com/microsoft/PowerToys/assets/22452178/49c81b74-e2e1-4ef6-9adb-ee7d422a7766)

htcfreek commented:
> Currently, the plugin doesn't show a list of supported GUID versions. It just says "Undefined GUID version", even for version 2, which is defined, but I didn't implement it here.
> But yes, I think it should show the supported versions and have a separate error message for version 2.
> 
> Perhaps like this?
> ![image](https://github.com/microsoft/PowerToys/assets/22452178/3ae7ac85-3eee-489a-b998-6a532f426024)
> ![image](https://github.com/microsoft/PowerToys/assets/22452178/49c81b74-e2e1-4ef6-9adb-ee7d422a7766)
> 

I would go with only the first (unsupported) error message. The difference between unsupported and undefeated ifs not important for the user and only a technical point.

It should be shown even for the query "guid" where the user missed the version. Or do we fallback to a specific version in this case?

IHorvalds commented:
If the user skipped the version it falls back to version 4. I think that's a sensible default since using random GUIDs seems like the most common use case and also doesn't require any parameters.

About the error message, I don't think "wrong version" is a good idea because it implies a "right" version. I also think the difference between unsupported and undefined is somewhat important. The RFC 4122 only defines the 5 versions, but there is apparently a [draft for some new ones](https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format-04#name-uuid-version-6).
I know it probably won't make a difference to most users, but we may end up implementing version 2 and at some point even the new versions.

htcfreek commented:
> If the user skipped the version it falls back to version 4. I think that's a sensible default since using random GUIDs seems like the most common use case and also doesn't require any parameters.

Agree. 👍

> About the error message, I don't think "wrong version" is a good idea because it implies a "right" version. I also think the difference between unsupported and undefined is somewhat important. The RFC 4122 only defines the 5 versions, but there is apparently a [draft for some new ones](https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format-04#name-uuid-version-6).
> I know it probably won't make a difference to most users, but we may end up implementing version 2 and at some point even the new versions.

But at the end all these version are not supported by the plugin. So let's have "Unsupported GUID version" for all cases to keep it simple.

IHorvalds committed:
Change error message for unsupported GUID versions

IHorvalds commented:
Done :)

htcfreek reviewed (commented):
Ready for the final review from the core team.

jaimecbernardo committed:
Remove Any CPU from the solution

jaimecbernardo commented:
Hi @IHorvalds ,
Thank you for contributing this feature! :)
I committed a quick fix to delete the "Any CPU" configuration that was added by this PR.
I plan to review it soon.

IHorvalds commented:
@jaimecbernardo Thank you! :)

jaimecbernardo committed:
Add to installer

jaimecbernardo commented:
I've added a commit to add the plugin to the installer. I'll also start  a release CI run to double check.

IHorvalds commented on installer/PowerToysSetup/Resources.wxs:
```diff
@@ -11,7 +11,7 @@
     <!-- Resource directories should be added only if the installer is built on the build farm -->
     <?ifdef env.IsPipeline?>
-      <?foreach ParentDirectory in LauncherInstallFolder;FancyZonesInstallFolder;ImageResizerInstallFolder;ColorPickerInstallFolder;FileExplorerPreviewInstallFolder;HistoryPluginFolder;CalculatorPluginFolder;FolderPluginFolder;ProgramPluginFolder;ShellPluginFolder;IndexerPluginFolder;UnitConverterPluginFolder;ValueGeneratorPluginFolder;UriPluginFolder;WindowWalkerPluginFolder;OneNotePluginFolder; ...
+      <?foreach ParentDirectory in LauncherInstallFolder;FancyZonesInstallFolder;ImageResizerInstallFolder;ColorPickerInstallFolder;FileExplorerPreviewInstallFolder;HistoryPluginFolder;CalculatorPluginFolder;FolderPluginFolder;ProgramPluginFolder;ShellPluginFolder;IndexerPluginFolder;UnitConverterPluginFolder;ValueGeneratorPluginFolder;UriPluginFolder;WindowWalkerPluginFolder;OneNotePluginFolder; ...
```
`ValueGeneratorPluginFolder` was already in this list (between UnitConverterPluginFolder and UriPluginFolder).

IHorvalds replied:
I thought I added the plugin to the installer a few commits ago, but it seems I missed a lot of things. Thank you for fixing that @jaimecbernardo!
Is there any documentation on how to do this so I can do this right before requesting a review next time? When I built it the powertoys solution and the installer locally the plugin was installed by default.

htcfreek replied:
We should order the lists alphabetical

jaimecbernardo replied:
Sorry, this indeed ended up failing release CI :) Thanks for noticing it.
Regarding documentation, we'd definitely benefit from it. But it's also a moving target, https://github.com/microsoft/PowerToys/pull/27451 will change it again, for example.

jaimecbernardo committed:
Remove duplicated ValueGeneratorPluginFolder entry

jaimecbernardo reviewed (approved):
LGTM! Thank you for the contribution!
It works well.

Only issue I have here is that it will be pretty hard to use without the documentation. Perhaps a more detailed description of the plugin in Settings or even a different default result when we're using the activation command would help here, but I think this can go in as it is.
Thank you, once again! ;)

IHorvalds commented:
Thank you! :)
